### PR TITLE
Equation String to AGraph Functionality

### PR DESCRIPTION
**Pull request by nightdr**
Added functionality to go to and from sympy representations of equations (including strings) to agraphs.

### CODE REVIEW

File: bingo/symbolic_regression/agraph/string_generation.py [hunk 1]
```diff
@@ -41,10 +43,24 @@
...
+                   LOGARITHM: "log({})",
+                   POWER: "({})**({})",
+                   ABS: "abs({})",
+                   SQRT: "sqrt({})",
+                   SAFE_POWER: "abs({})**({})"}
 CONSOLE_PRINT_MAP = {ADDITION: "{} + {}",
                      SUBTRACTION: "{} - ({})",
                      MULTIPLICATION: "({})({})",
-                     DIVISION: "({})/({}) ",
+                     DIVISION: "({})/({}) ",  # TODO extra space intentional?
```
Comment (nightdr) [inline bingo/symbolic_regression/agraph/string_generation.py:63]: I was curious if this extra space was intentional

File: bingo/symbolic_regression/agraph/agraph.py [hunk 2]
```diff
@@ -104,22 +114,42 @@ class AGraph(Equation, continuous_local_opt.ChromosomeInterface):
...
+            else:  # is str instance
+                self._sympy_expr = parse_expr(sympy_representation)
+                self._sympy_str = str(self._sympy_expr)
+                # not using sympy_representation directly for self._sympy_str
+                # because it might not be in simplified form
+            command_array, constants = \
+                sympy_string_to_command_array_and_constants(self._sympy_str)
+            self.set_local_optimization_params(constants)
+            self.command_array = command_array
+            # TODO decide on setting _needs_opt or not
```
Comment (nightdr) [inline bingo/symbolic_regression/agraph/agraph.py:147]: A couple of considerations for setting up the AGraph from a Sympy string. If we get a string like "1.0 + 2.0", we get an AGraph that adds two constants together and sets those constants at 1.0 and 2.0 respectively. Should _needs_opt be true in this case or should it be false like normal and be left for the user to decide?

File: bingo/symbolic_regression/agraph/agraph.py [hunk 3]
```diff
@@ -104,22 +114,42 @@ class AGraph(Equation, continuous_local_opt.ChromosomeInterface):
...
+            # TODO how to deal with command_array being updated after this?
```
Comment (nightdr) [inline bingo/symbolic_regression/agraph/agraph.py:148]: There's also the issue of if we want to keep the sympy_expr and sympy_str attributes updated as the agraph updates

File: bingo/symbolic_regression/agraph/string_generation.py (hunk 1 above)
Comment (gbomarito) [reply bingo/symbolic_regression/agraph/string_generation.py:63]: not intentional, this is a typo

**Review by gbomarito — CHANGES_REQUESTED** [review]
Thanks!

File: bingo/symbolic_regression/agraph/string_parsing.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    while len(stack) > 0:
+        token = stack.pop()
+        if token == "(":
+            raise RuntimeError("Mismatched parenthesis")
+        output.append(token)
+
+    return output
+
+
+def postfix_to_command_array_and_constants(postfix_tokens):
```
[the remark this answers is not in the document]
Comment (nightdr) [reply bingo/symbolic_regression/agraph/string_parsing.py:69]: sounds good to me. I added this in

File: bingo/symbolic_regression/agraph/agraph.py (hunk 2 above)
Comment (nightdr) [reply bingo/symbolic_regression/agraph/agraph.py]: agreed, made needs_opt = True the default in the new changes

File: bingo/symbolic_regression/agraph/string_generation.py (hunk 1 above)
Comment (nightdr) [reply bingo/symbolic_regression/agraph/string_generation.py]: sounds good, removed in new changes

**Comment by nightdr** [comment]
rebased for cleaner working tree

**Pull request merged**